internal/graph/iterator_stream.go

```diff
@@ -0,0 +1,133 @@
+package graph
+
+import (
+	"context"
+	"slices"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+type iteratorStream struct {
+	idx    int
+	buffer storage.TupleKeyIterator
+	done   bool
+	source chan *iteratorMsg
+}
+
+// helper function to drain the stream until the head's object is larger than target.
+// If the buffer is drained and no more items, it will set to stop and buffer will be nil.
+func (s *iteratorStream) skipToTarget(ctx context.Context, target string) error {
+	t, err := s.buffer.Head(ctx)
+	if err != nil {
+		if storage.IterIsDoneOrCancelled(err) {
+			s.buffer.Stop()
+			s.buffer = nil
+			return nil
+		}
+		return err
+	}
+	tmpKey := t.GetObject()
+	for tmpKey < target {
+		_, _ = s.buffer.Next(ctx)
+		t, err = s.buffer.Head(ctx)
+		if err != nil {
+			if storage.IterIsDoneOrCancelled(err) {
+				s.buffer.Stop()
+				s.buffer = nil
+				break
+			}
+			return err
+		}
+		tmpKey = t.GetObject()
+	}
+	return nil
+}
+
+// drain all item in the stream's buffer and return these items.
+func (s *iteratorStream) drain(ctx context.Context) ([]*openfgav1.TupleKey, error) {
+	var batch []*openfgav1.TupleKey
+	for {
+		t, err := s.buffer.Next(ctx)
+		if err != nil {
+			if storage.IterIsDoneOrCancelled(err) {
+				s.buffer.Stop()
+				s.buffer = nil
+				break
+			}
+			return nil, err
+		}
+		batch = append(batch, t)
+	}
+	return batch, nil
+}
+
+// for all slices with index in itersWithEqualObject, call next and return the first item
+// return error if any individual stream in the streamSlices is empty.
```

Comment:
This godoc makes little sense to me

- `for all slices with index`? what index?
- what is `itersWithEqualObject` (looks like it's `streamToProcess`)
- `call next and return the first item` but you are overwriting the `item` variable on every loop, so presumably it actually returns the _last_ item?